app/com/yetu/oauth2provider/oauth2/services/AuthorizeService.scala

```diff
@@ -44,55 +46,39 @@ class AuthorizeErrorHandler(clientService: IClientService,
+    clientService
+      .findClient(request.clientId)
+      .map(c => try {
 
-    if (!client.coreYetuClient) {
-      scopeService.getScopeFromPermission(
-        permissionService.findPermission(user.identityId.userId, client.clientId))
-    }
+        val client = c.getOrElse(throw new InvalidClient(s"client_id '${request.clientId}' does not exist"))
+        val validRedirectUrls = client.redirectURIs
+        val redirectUrl = URLDecoder.decode(request.redirectUri, "UTF-8")
 
-    request.scope.foreach { scope =>
-      scope.split(' ').toList.foreach { requestScope =>
-        if (!validScopes.contains(requestScope)) {
-          throw new InvalidScope(s"invalid scope: $requestScope")
-        }
-      }
-    }
+        if (!validRedirectUrls.contains(redirectUrl)) {
 
-    val validRedirectUrls = client.redirectURIs
-    val redirectUrl = URLDecoder.decode(request.redirectUri, "UTF-8")
+          logger.warn(s"clientID:[${client.clientId}] request redirect url is NOT VALID! " +
+            s"[$redirectUrl]. Only allowed ones are : $validRedirectUrls}")
 
-    if (!validRedirectUrls.contains(redirectUrl)) {
-
-      logger.warn(s"clientID:[${client.clientId}] request redirect url is NOT VALID! " +
-        s"[$redirectUrl]. Only allowed ones are : $validRedirectUrls}")
+          if (Config.redirectURICheckingEnabled) {
+            throw new RedirectUriMismatch(s"invalid redirect url.")
+          }
+        }
 
-      if (Config.redirectURICheckingEnabled) {
-        throw new RedirectUriMismatch(s"invalid redirect url.")
-      }
-    }
+        val authorizedClient = AuthorizedClient(client, request, redirectUrl)
+        Right(authorizedClient)
 
-    val authorizedClient = AuthorizedClient(client, request, redirectUrl)
-    Right(authorizedClient)
+      } catch {
+        case e: OAuthError => Left(e)
+      })
 
   } catch {
-    case e: OAuthError => Left(e)
+    case e: OAuthError => Future.successful(Left(e))
   }
 
-  def validateParameters[A](user: YetuUser)(callback: AuthorizedClient => Result)(implicit request: play.api.mvc.Request[A]): Result = {
-    handleAuthorizeRequest(request, user) match {
-      case Left(e) if e.statusCode == 400 => BadRequest.withHeaders(responseOAuthErrorHeader(e))
-      case Left(e) if e.statusCode == 401 => Unauthorized.withHeaders(responseOAuthErrorHeader(e))
-      case Right(client)                  => callback(client)
-    }
-  }
-
-  def validateParametersAsync[A](user: YetuUser)(callback: AuthorizedClient => Future[Result])(implicit request: play.api.mvc.Request[A]): Future[Result] = {
-    handleAuthorizeRequest(request, user) match {
+  def validateParameters[A](user: YetuUser)(callback: AuthorizedClient => Future[Result])(implicit request: play.api.mvc.Request[A]): Future[Result] = {
+    handleAuthorizeRequest(request, user).flatMap {
       case Left(e) if e.statusCode == 400 => Future.successful(BadRequest.withHeaders(responseOAuthErrorHeader(e)))
```

Comment:
Doesn't Play allow you to map exceptions to result codes? I would use a failed `Future` for the error cases and then have the error management in a centralized place